docs/source/Library.md

```diff
@@ -33,179 +39,196 @@ Next we specify the core model itself. Here we will build a small model with an
                                              word_padding_idx=src_padding)
 
 encoder = onmt.encoders.RNNEncoder(hidden_size=rnn_size, num_layers=1,
-                                 rnn_type="LSTM", bidirectional=True,
-                                 embeddings=encoder_embeddings)
+                                   rnn_type="LSTM", bidirectional=True,
+                                   embeddings=encoder_embeddings)
 
-decoder_embeddings = onmt.modules.Embeddings(emb_size, len(vocab["tgt"]),
+decoder_embeddings = onmt.modules.Embeddings(emb_size, len(tgt_vocab),
                                              word_padding_idx=tgt_padding)
-decoder = onmt.decoders.decoder.InputFeedRNNDecoder(hidden_size=rnn_size, num_layers=1,
-                                           bidirectional_encoder=True,
-                                           rnn_type="LSTM", embeddings=decoder_embeddings)
+decoder = onmt.decoders.decoder.InputFeedRNNDecoder(
+    hidden_size=rnn_size, num_layers=1, bidirectional_encoder=True, 
+    rnn_type="LSTM", embeddings=decoder_embeddings)
+
 model = onmt.models.model.NMTModel(encoder, decoder)
 
 # Specify the tgt word generator and loss computation module
 model.generator = nn.Sequential(
-            nn.Linear(rnn_size, len(vocab["tgt"])),
-            nn.LogSoftmax())
-loss = onmt.utils.loss.NMTLossCompute(model.generator, vocab["tgt"])
+    nn.Linear(rnn_size, len(tgt_vocab)),
+    nn.LogSoftmax())
+
+loss = onmt.utils.loss.NMTLossCompute(
+    criterion=nn.NLLLoss(ignore_index=tgt_padding, reduction='sum'),
+    generator=model.generator)
 ```
 
-Now we set up the optimizer. This could be a core torch optim class, or our wrapper which handles learning rate updates and gradient normalization automatically.
+Now we set up the optimizer. Our wrapper around a core torch optim class handles learning rate updates and gradient normalization automatically.
 
 
 ```python
-optim = onmt.utils.optimizers.Optimizer(method="sgd", learning_rate=1, max_grad_norm=2)
-optim.set_parameters(model.named_parameters())
+lr = 1
+torch_optimizer = torch.optim.SGD(model.parameters(), lr=lr)
+optim = onmt.utils.optimizers.Optimizer(
+    torch_optimizer, learning_rate=lr, max_grad_norm=2)
 ```
 
-Now we load the data from disk. Currently will need to call a function to load the fields into the data as well.
+Now we load the data from disk with the values of the associated vocab fields. To iterate through the data itself we use a wrapper around a torchtext iterator class. We specify one for both the training and test data.
 
 
 ```python
 # Load some data
-data = torch.load("../../data/data.train.1.pt")
-valid_data = torch.load("../../data/data.valid.1.pt")
-data.load_fields(vocab)
-valid_data.load_fields(vocab)
-data.examples = data.examples[:100]
+from itertools import chain
+train_data_file = "data/data.train.0.pt"
+valid_data_file = "data/data.valid.0.pt"
+dataset_fields = dict(chain.from_iterable(vocab_fields.values()))
```

Comment:
It is not very intuitive that iterators expect "fields" to be a dict[str, List[Tuple[str, Field]]] whereas Tranlator or TranslatorBuilder expect "fields" to be a dict[str, list[tuple[str, torchtext.data.Field]]].
Maybe choose discriminant names?

Reply:
Or they should all expect the same structure. See this comment

https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/train_single.py#L139

The docs for the format that DatasetLazyIter expects is evidently wrong.

Reply:
Should they? As for now, giving DatasetLazyIter a dict[str, torchtext.data.Field] actually makes the code work, while giving it the original dict[str, list[tuple[str, torchtext.data.Field]]] does not allow training to start because the structure is not correct. Could you expand a bit more on the logic behind?

Reply:
You're correct that DatasetLazyIter actually needs dict[str, torchtext.data.Field]. The docs claim DatasetLazyIter expects dict[str, list[tuple[str, torchtext.data.Field]]] (I wrote those docs and I got it wrong because I incorrectly assumed it would take the same structure as the other ``field`` arguments, sorry!).

Now, the comment in ``train_single.py`` that I linked above says that ``DatasetLazyIter`` expecting a different structure is (or, should have been) a temporary thing. It's a trivial fix. Move the ``dict(itertools.chain.from_iterable(fields.values()))`` business inside ``DatasetLazyIter``.

But, it gets weird because when torchtext says ``fields``, e.g. in [Dataset](https://github.com/pytorch/text/blob/master/torchtext/data/dataset.py#L34), they're talking about ``list[tuple[str, Field]]``. Onmt uses the extra level of nesting because onmt text features (POS tags, NER, etc) used to be handled as separate fields. Now, that's unnecessary because of ``TextMultiField``.

All to say, yes they should be named different and the docs should be updated OR they should expect the same thing (either the extra level of nesting or, more elegantly, without it so that it matches torchtext).

Reply:
Well, torchtext is inconsistent about what it means too:
https://github.com/pytorch/text/blob/master/torchtext/data/dataset.py#L23 (fields is a dict[str, Field])
https://github.com/pytorch/text/blob/master/torchtext/data/example.py#L19 (fields should be dict[str, list[tuple[str, Field]]] or dict[str, tuple[str, Field]])

Reply:
Interesting, I let it that way at the moment in the example, just expliciting a bit more, let me know when things are set on how onmt handles this.